fendermint/vm/interpreter/src/fvm/state/snapshot.rs

```diff
@@ -0,0 +1,364 @@
+        state_params: FvmStateParams,
+        block_height: BlockHeight,
+    ) -> anyhow::Result<Self> {
+        let state_tree =
+            StateTree::new_from_root(ReadOnlyBlockstore::new(store), &state_params.state_root)?;
+
+        Ok(Self {
+            state_tree,
+            state_params,
+            block_height,
+        })
+    }
+
+    fn from_root(store: DB, root_cid: Cid) -> anyhow::Result<Self> {
+        if let Some((state_params, block_height)) = store.get_cbor::<BlockStateParams>(&root_cid)? {
+            let state_tree_root = state_params.state_root;
+            Ok(Self {
+                state_tree: StateTree::new_from_root(
+                    ReadOnlyBlockstore::new(store),
+                    &state_tree_root,
+                )?,
+                state_params,
+                block_height,
+            })
+        } else {
+            Err(anyhow!(
+                "invalid v1 snapshot, root cid not found: {}",
+                root_cid
+            ))
+        }
+    }
+
+    fn into_streamer(self) -> anyhow::Result<(Cid, SnapshotStreamer)> {
+        let state_tree_root = self.state_params.state_root;
+
+        let block_state_params = (self.state_params, self.block_height);
+        let bytes = fvm_ipld_encoding::to_vec(&block_state_params)?;
+        let root_cid = Cid::new_v1(DAG_CBOR, Code::Blake2b256.digest(&bytes));
+
+        let state_tree_streamer =
+            StateTreeStreamer::new(state_tree_root, self.state_tree.into_store());
+        let root_streamer = tokio_stream::iter(vec![(root_cid, bytes)]);
+        let streamer: SnapshotStreamer = Box::new(state_tree_streamer.merge(root_streamer));
+
+        Ok((root_cid, streamer))
+    }
+}
+
+#[pin_project::pin_project]
+struct StateTreeStreamer<BlockStore> {
+    /// The list of cids to pull from the blockstore
+    #[pin]
+    dfs: VecDeque<Cid>,
+    /// The block store
+    bs: BlockStore,
+}
+
+impl<BlockStore> StateTreeStreamer<BlockStore> {
+    pub fn new(state_root_cid: Cid, bs: BlockStore) -> Self {
+        let mut dfs = VecDeque::new();
+        dfs.push_back(state_root_cid);
+        Self { dfs, bs }
+    }
+}
+
+impl<BlockStore: Blockstore> Stream for StateTreeStreamer<BlockStore> {
+    type Item = (Cid, Vec<u8>);
+
+    fn poll_next(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+
+        loop {
+            let cid = if let Some(cid) = this.dfs.pop_front() {
+                cid
+            } else {
+                return Poll::Ready(None);
+            };
+
+            match this.bs.get(&cid) {
+                Ok(Some(bytes)) => {
+                    let ipld = from_slice::<Ipld>(&bytes).unwrap();
+                    walk_ipld_cids(ipld, &mut this.dfs);
+                    return Poll::Ready(Some((cid, bytes)));
+                }
+                Ok(None) => {
+                    tracing::warn!("cid: {cid:?} has no value in block store, skip");
```

Comment:
```suggestion
                    tracing::debug!("cid: {cid:?} has no value in block store, skip");
```
I think these will be part of the data model, some CIDs are just that without any reachable value in the database, but this is nothing the operator can do anything about. I would reserve `warn` to things they should keep their eyes on, and demote this to `debug` for times when you are wondering what is going on. There could be unpredictable numbers of these, so we don't want the logs to start sending emails to operators by the thousands every time a snapshot is made.

Reply:
updated